File: backend/src/main/java/com/example/subscriptions/service/SubscriptionService.java
package com.example.subscriptions.service;

import java.time.LocalDate;
import java.util.List;

import org.springframework.stereotype.Service;

import com.example.subscriptions.dao.SubscriptionDAO;
import com.example.subscriptions.model.Subscription;

@Service
public class SubscriptionService {

    private final SubscriptionDAO subscriptionDAO;

    public SubscriptionService(SubscriptionDAO subscriptionDAO) {
        this.subscriptionDAO = subscriptionDAO;
    }

    public List<Subscription> getAllSubscriptions() {
        return subscriptionDAO.findAll();
    }

    public Subscription getSubscriptionById(Long id) {
        return subscriptionDAO.findById(id);
    }

    public void addSubscription(Subscription subscription) {
        subscriptionDAO.save(subscription);
    }

    public void updateSubscription(Long id, Subscription subscription) {
        subscriptionDAO.update(id, subscription);
    }

    public void deleteSubscription(Long id) {
        subscriptionDAO.deleteById(id);
    }

    public List<Subscription> searchSubscriptions(String channelName, String channelUrl, LocalDate subscribedDate) {
        return subscriptionDAO.search(channelName, channelUrl, subscribedDate);
    }
}

File: frontend/src/app/subscription.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Subscription } from './subscription';

@Injectable({
  providedIn: 'root'
})
export class SubscriptionService {
  private baseUrl = 'http://localhost:8080/api/subscriptions';

  constructor(private http: HttpClient) {}

  getSubscriptions(): Observable<Subscription[]> {
    return this.http.get<Subscription[]>(`${this.baseUrl}`);
  }

  getSubscriptionById(id: number): Observable<Subscription> {
    return this.http.get<Subscription>(`${this.baseUrl}/${id}`);
  }

  addSubscription(subscription: Subscription): Observable<Subscription> {
    return this.http.post<Subscription>(`${this.baseUrl}`, subscription);
  }

  updateSubscription(id: number, subscription: Subscription): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/${id}`, subscription);
  }

  deleteSubscription(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }


  searchSubscriptions(channelName?: string, channelUrl?: string, subscribedDate?: string): Observable<Subscription[]> {
    let params = new HttpParams();

    if (channelName) {
      params = params.set('channelName', channelName);
    }
    if (channelUrl) {
      params = params.set('channelUrl', channelUrl);
    }
    if (subscribedDate) {
      params = params.set('subscribedDate', subscribedDate);
    }

    return this.http.get<Subscription[]>(`${this.baseUrl}/search`, { params });
  }

  
}
